Add all of the builtins

disconcision opened the pull request:
1. What: This PR add a 3 types and 74 functions or function alises to the builtins, and reorganizes the existing builtins to streamline the process of adding new ones. The functions that could be standard library functions are implemented as Hazel code, with accompanying concrete syntax.
2. Why: This greatly expanded selection is intended as a temporary measure to support LLM, MVU, dataframes, and advent of code initiatives, and is to later be replaced by a Hazel standard library.
3. Non-total functions return 'undefined' as the dynamic error constructor doesn't seem to show up in the value display. All added functions have dynamic tests.

 Types:
 - Option = None + Some(?) (used for several list functions)
 - Either = Left(?) + Right(?) (used for partition_map)
 - Ord = Lt + Eq + Gt (used for sort)

 String functions (some use regexps provided as plain strings)
 - string_escaped
 - string_uppercase
 - string_lowercase
 - string_capitalize
 - string_uncapitalize
 - string_match
 - string_replace
 - string_search

 Pair functions:
 - fst
 - snd

 Option functions:
 - option_map
 - option_bind
 - option_to_list

 List functions:
 - cons
 - hd
 - tl
 - is_empty
 - length
 - map
 - filter
 - fold_left
 - fold_right
 - flat_map
 - zip
 - unzip
 - reverse
 - take
 - drop
 - take_while
 - drop_while
 - slice
 - init
 - range
 - enumerate
 - any
 - all
 - intersperse
 - nth
 - append
 - concat
 - mapi
 - filteri
 - find
 - mem
 - partition
 - rev_append
 - fold_left2
 - fold_right2
 - map2
 - all2
 - any2

 List functions involving option:
 - hd_opt
 - tl_opt
 - filter_map
 - nth_opt
 - find_opt
 - find_index
 - find_map
 - find_mapi

 List functions involving either:
 - partition_map

 Association list functions:
 - assoc
 - assoc_opt
 - mem_assoc
 - remove_assoc

 List function aliases:
 - contains
 - flatten
 - rev
 - exists
 - for_all
 - exists2
 - for_all2
 - concat_map
 - split
 - combine

7h3kk1d committed:
Add HazelFn support in builtins

- demo for @disconcision

7h3kk1d committed:
Add 'map' built-in function implementation

7h3kk1d committed:
ought to be enough for anybody

7h3kk1d committed:
Add built-in functions: reverse, take, drop, contains, range, enumerate, any, all, intersperse

disconcision committed:
more list functions + tests

7h3kk1d committed:
Fix tests

- Fix incorrect usage of builtins for builtin constructors
- Add correct type to constructors in both tests and implementation

7h3kk1d committed:
Rewrite Test_Evaluator_BuiltinsLists to use parse_and_evaluate_test

- Remove all negative ints from assertions because of issues with `UnOp(Minus, int)` and Int(negative)`
- Modify TermBase.fast_equal to have an ignore constructor type  parameter

disconcision committed:
builtin option and result types. options fns

disconcision committed:
fix option builtins. more list builtins

disconcision committed:
more builtin string functions included regexp match, replace, search

disconcision committed:
tests for new string builtins

disconcision committed:
sort debugging

disconcision committed:
fix sort. tests are skipped but will pass after #1729

disconcision committed:
cleanup builtins dir structure

disconcision committed:
builtins cleanup

disconcision committed:
builtins cleanup

disconcision committed:
accidentally a file

7h3kk1d commented on test/Test_Menhir.re:
```diff
@@ -14,11 +14,12 @@ let strip_Wrap_and_add_builtins =
         | Probe(e, _) => cont(e)
         | Var(x) =>
-          let builtin = Util.VarMap.lookup(Builtins.Pervasives.builtins, x);
+          let builtin = Util.VarMap.lookup(Builtins.builtin_map, x);
```
So it looks like some of the tests are breaking since the documentation buffers are shadowing the builtins (map).  We do the builtins based on the context which is shadowable and not in the parser ordinarily.

I think we should just stop parsing builtins as builtins in menhir. The original use case to mostly use it for elaboration tests has faded somewhat and the only test that fails if we parse them as vars is `Built-in function (menhir).`. If we end up needing a literal syntax for the builtins (which I hope we don't) we can add it back to menhir with special syntax.

7h3kk1d replied:
Let me know if you want me to make this change

disconcision replied:
@7h3kk1d yeah i was going to ask you why the parser was handling builtins specially but i forgot. i think we should get rid of that if we can. if you could do it that would be great

disconcision replied:
wait, what are the failing tests? are these ones that don't run with `make test-quick`?

disconcision replied:
oh. i see. i though test-quick just did the same tests but... faster?

7h3kk1d replied:
If only. It skips the pbt and the longer tylr ones

7h3kk1d commented on src/language/builtins/BuiltinsADT.re:
```diff
@@ -0,0 +1,192 @@
+
+// Convert type aliases to context entries
+let types: list(Ctx.entry) =
+  List.map(((name, typ)) => create_type_alias(name, typ), type_aliases);
```
It kills me that this doesn't uncurry automatically to make it point-free

disconcision replied:
sometimes i make functions take pairs for this reason, usually makes me feel somewhat dirty

7h3kk1d commented on src/language/builtins/BuiltinsBase.re:
```diff
@@ -0,0 +1,470 @@
+    ret: Atom(String),
+    imp: d => {
+      let-unbox s = (Atom(String), d);
+      Some(Exp.string(String.escaped(s)));
```
Do we want unescaped?

disconcision replied:
huh. i'll add it. missed it as it's not in the String lib (its in Scanf)

7h3kk1d commented on src/language/builtins/BuiltinsList.re:
```diff
@@ -0,0 +1,2844 @@
+  {
+    str: {|fix filter -> fun xs f -> case xs
+             | [] => []
+             | [x] => if f(x) then [x] else []
```
This middle case is unnecessary. Can't tell if it's here as an optimization.

7h3kk1d replied:
fold_left does the same thing

7h3kk1d commented on src/language/builtins/BuiltinsList.re:
```diff
@@ -0,0 +1,2844 @@
+  },
+  {
+    str: {|fix flat_map -> fun (xs, f) -> fold_left(xs, fun (acc,x) -> acc @ f(x), []))([1,2],fun x ->[x,x])|},
+    name: "flat_map", //TODO: weird string content above
```
maybe todo should be todone

7h3kk1d commented on src/language/builtins/BuiltinsList.re:
```diff
@@ -0,0 +1,2844 @@
+    },
+  },
+  {
+    str: {|fix reverse -> fun xs -> case xs
+             | [] => []
+             | x :: xs => reverse(xs) @ [x]
+           end|},
```
I expected this to be slow and was going to suggest

```hazel
let reverse = let go = fun (xs, acc) -> case xs 
| [] => acc
| (x :: xs) => go(xs, x::acc) 
end in
go(_, []) 
```

But since list concatenation is implemented in ocaml the existing implementation is way faster. I still suspect it's quadratic but thousands of times faster.

7h3kk1d commented on src/language/builtins/BuiltinsList.re:
```diff
@@ -0,0 +1,2844 @@
+             | [] => []
+             | _ => if n <= 0 then [] else case xs
+             | x :: xs => x :: take(xs, n - 1)
+           end|},
```
For take and drop if you put the if on the outside you don't need the double case statement which also would make the matches exhaustive.

7h3kk1d commented on src/language/builtins/BuiltinsList.re:
```diff
@@ -0,0 +1,2844 @@
+    },
+  },
+  {
+    str: {|fix enumerate -> fun xs -> enumerate_helper(xs, 0)
+           fix enumerate_helper -> fun (xs, index) -> case xs
+             | [] => []
+             | x :: xs => (index, x) :: enumerate_helper(xs, index + 1)
+           end|},
```
Fwiw, this string isn't accurate.

7h3kk1d commented on src/language/builtins/BuiltinsList.re:
```diff
@@ -0,0 +1,2844 @@
+    },
+  },
+  {
+    str: {|fix mem -> fun (xs, x) -> case xs
+             | [] => false
+             | y :: xs => if x == y then true else mem(xs, x)
+           end|},
```
Where we can I'd probably start using the stdlib

```hazel
fun xs, x -> any(xs, fun y -> x == y)
```

7h3kk1d commented on src/language/builtins/BuiltinsList.re:
```diff
@@ -0,0 +1,2844 @@
+    },
+  },
+  {
+    str: {|fix slice -> fun (start, len, xs) -> take(drop(xs, start),len)|},
```
We should consider if we want end or length on slice. Python/js do end index but I could see it either way.

disconcision replied:
changed it

7h3kk1d commented on src/language/term/TermBase.re:
```diff
@@ -177,8 +177,8 @@ and Exp: {
 
-  let fast_equal: (t, t) => bool;
-  let equal: (t, t) => bool;
+  let fast_equal: (~ignore_constructor_types: bool=?, t, t) => bool;
```
Self reminder to do https://github.com/hazelgrove/hazel/issues/1563

7h3kk1d commented on test/evaluator/Test_Evaluator_Prelude.re:
```diff
@@ -3,7 +3,11 @@ open Language;
+let dhexp_typ =
+  testable(
+    Fmt.using(Exp.show, Fmt.string),
+    DHExp.fast_equal(~ignore_constructor_types=true) // This is a bad idea to just blanket ignore the type
```
Are we good just keeping this for now? I'm not sure where the constructor types are used in dynamics and when it would matter.

disconcision replied:
i didn't know we had types on constructors sooo

7h3kk1d replied:
types on types on types

disconcision replied:
is this actually necessary? didn't we change it so that the tests do have the types on them?

7h3kk1d replied:
These tests will fail:

https://github.com/hazelgrove/hazel/pull/1734/files#diff-a5b6e71606758826f0a3f4c4052b3fa0223d021509a0238b4103574a0127d484R209

But we can rewrite those not to use the parser for just the ones that have constructors.

disconcision replied:
This is the only remaining issue; are you okay just leaving this in place for now?

7h3kk1d committed:
Remove built-in function support from Menhir parser

This commit removes the support for built-in functions from the Menhir parser and lexer, including related tokens and parsing rules. It also updates the test files to eliminate references and test cases for built-in functions, simplifying the parser logic and focusing on core language features. Additionally, it adjusts the expression mapping to no longer handle built-in functions as special cases.

disconcision committed:
cleanup/minor adjustments as per pr #1734

disconcision committed:
Remove built-in function support from Menhir parser (#1736)

disconcision committed:
builtin cleanup

disconcision committed:
Merge branch 'dev' of github.com:hazelgrove/hazel into demo-hzl-builtins-moar

7h3kk1d committed:
Add ignore_constructor_types flag to tests

Explicitly opt out of the constructor type checking
